pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/StringDictionary.java

```diff
@@ -41,84 +47,117 @@ public DataType getValueType() {
-    return BytesUtils.toBytes(getUnpaddedString(dictId, getBuffer()));
+    return BytesUtils.toBytes(internStringValue(dictId, getBuffer()));
   }
 
   @Override
   public void readIntValues(int[] dictIds, int length, int[] outValues) {
     byte[] buffer = getBuffer();
     for (int i = 0; i < length; i++) {
-      outValues[i] = Integer.parseInt(getUnpaddedString(dictIds[i], buffer));
+      outValues[i] = Integer.parseInt(internStringValue(dictIds[i], buffer));
     }
   }
 
   @Override
   public void readLongValues(int[] dictIds, int length, long[] outValues) {
     byte[] buffer = getBuffer();
     for (int i = 0; i < length; i++) {
-      outValues[i] = Long.parseLong(getUnpaddedString(dictIds[i], buffer));
+      outValues[i] = Long.parseLong(internStringValue(dictIds[i], buffer));
     }
   }
 
   @Override
   public void readFloatValues(int[] dictIds, int length, float[] outValues) {
     byte[] buffer = getBuffer();
     for (int i = 0; i < length; i++) {
-      outValues[i] = Float.parseFloat(getUnpaddedString(dictIds[i], buffer));
+      outValues[i] = Float.parseFloat(internStringValue(dictIds[i], buffer));
     }
   }
 
   @Override
   public void readDoubleValues(int[] dictIds, int length, double[] outValues) {
     byte[] buffer = getBuffer();
     for (int i = 0; i < length; i++) {
-      outValues[i] = Double.parseDouble(getUnpaddedString(dictIds[i], buffer));
+      outValues[i] = Double.parseDouble(internStringValue(dictIds[i], buffer));
     }
   }
 
   @Override
   public void readStringValues(int[] dictIds, int length, String[] outValues) {
     byte[] buffer = getBuffer();
     for (int i = 0; i < length; i++) {
-      outValues[i] = getUnpaddedString(dictIds[i], buffer);
+      outValues[i] = internStringValue(dictIds[i], buffer);
     }
   }
 
   @Override
   public void readBytesValues(int[] dictIds, int length, byte[][] outValues) {
     byte[] buffer = getBuffer();
     for (int i = 0; i < length; i++) {
-      outValues[i] = BytesUtils.toBytes(getUnpaddedString(dictIds[i], buffer));
+      outValues[i] = BytesUtils.toBytes(internStringValue(dictIds[i], buffer));
+    }
+  }
+
+  private String internStringValue(int dictId) {
+    if (_internTable == null) {
+      return getUnpaddedString(dictId, getBuffer());
+    }
+    String interned = _internTable[dictId];
+    if (interned == null) {
+      interned = getUnpaddedString(dictId, getBuffer());
+      _internTable[dictId] = interned;
+    }
+    return interned;
+  }
+
+  private String internStringValue(int dictId, byte[] buffer) {
+    if (_internTable == null) {
+      return getUnpaddedString(dictId, buffer);
+    }
+    String interned = _internTable[dictId];
+    if (interned == null) {
+      interned = getUnpaddedString(dictId, buffer);
+      _internTable[dictId] = interned;
+    }
+    return interned;
+  }
+
+  @Override
+  public void close()
+      throws IOException {
+    if (_internTable != null) {
+      Arrays.fill(_internTable, null);
```

Comment:
Is this required? If so, does setting `_internTable = null` have better performance?

Reply:
I don’t think we need this, but note that the intern table is final. I’ll just remove it.